libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/impl/MapLibrePrimaryRouteDrawer.java

```diff
@@ -0,0 +1,340 @@
+            return;
+        }
+
+        if (lastLocationPoint == null) {
+            lastLocationPoint = Point.fromLngLat(location.getLongitude(), location.getLatitude());
+            return;
+        }
+
+        if (route == null) {
+            return;
+        }
+
+        String routeGeometry = route.geometry();
+        if (routeGeometry == null) {
+            //TODO: when this can happen?!
+            return;
+        }
+
+        if (valueAnimator != null) {
+            valueAnimator.cancel();
+        }
+
+        valueAnimator = ValueAnimator.ofFloat(0f, 1.0f);
+        valueAnimator.setInterpolator(new LinearInterpolator());
+
+        Point startPoint = lastLocationPoint;
+        Point targetPoint = Point.fromLngLat(location.getLongitude(), location.getLatitude());
+
+        // Location not changed, nothing to process
+        if (startPoint.equals(targetPoint)) {
+            return;
+        }
+
+        LineString routeLine = LineString.fromPolyline(routeGeometry, Constants.PRECISION_6);
+        LineString animatedSegment = TurfMisc.lineSlice(startPoint, targetPoint, routeLine);
+        double segmentLength = TurfMeasurement.length(animatedSegment, TurfConstants.UNIT_METERS);
+
+        valueAnimator.addUpdateListener(animation -> {
+            if (System.nanoTime() - lastUpdateTime < minUpdateIntervalNanoSeconds) {
+                return;
+            }
+
+            if (route == null) {
+                return;
+            }
+
+            float fraction = animation.getAnimatedFraction();
+            Point animatedPoint = TurfMeasurement.along(animatedSegment, segmentLength * fraction, TurfConstants.UNIT_METERS);
+            lastLocationPoint = animatedPoint;
+
+            drawRoute(route, animatedPoint, routeProgress);
+            lastUpdateTime = System.nanoTime();
+        });
+
+        valueAnimator.addListener(new AnimatorListenerAdapter() {
+            boolean isCanceled = false;
+
+            @Override
+            public void onAnimationCancel(Animator animation) {
+                isCanceled = true;
+            }
+
+            @Override
+            public void onAnimationEnd(Animator animation) {
+                if (isCanceled || route == null) {
+                    return;
+                }
+
+                lastLocationPoint = (Point) TurfMisc.nearestPointOnLine(targetPoint, routeLine.coordinates(), TurfConstants.UNIT_METERS).geometry();
+                drawRoute(route, lastLocationPoint, routeProgress);
+            }
+        });
+
+        long previousUpdateTimeStamp = locationUpdateTimestamp;
+        locationUpdateTimestamp = SystemClock.elapsedRealtime();
+
+        /* make animation slightly longer (with 1.1f) */
+        long duration = (long) ((locationUpdateTimestamp - previousUpdateTimeStamp) * 1.1);
+        duration = Math.min(duration, 2000L); //TODO: outsource max duration as field
+        valueAnimator.setDuration(duration);
+        valueAnimator.start();
```

Comment:
I think it is incredibly impressive what you've done here with the valueAnimator ❤️. 

I am wondering about the performance implication though. Have you tried and/or measured the energy impact of this, for example for a 100 or 1000km route?

For example nearest point on line can be quite costly for larger geometries.

Reply:
> Do you know how the source/feature-drawing is worked in detail? I have thought about to split the line in multiple features and have only update the active part, instead update the whole line. But I'm not sure if this will improve or degrade the performance.

I think this might be a good idea, so we could split the route line into geometries for every maneuver or leg. I think this will make other things more complicated though (like matching the features, to the current feature and only changing that etc.).

> Or maybe it make sense to draw the full line as driven and overlay with undriven path (or vice versa). In this case we need only calculate one line and also update only one feature.

This is actually an interesting idea as well. I think, as long as we stay on the Java layer, our options are somewhat limited.

@nyurik asked about this in the last navigation meeting. One more thing that we could consider, and I believe the maplibre-org would also consider adding to the maplibre-native, would be to actually enable this for maplibre-native. For example iOS is facing similar issues and I believe the Java implementation will never be as performant as a native solution. Maybe @louwers has some insight into this?

So my idea on a high level would be to allow a line to have different styles for different parts of the line, so maybe the first 50% has style A and the second 50% has style B. This might be useful for quite a lot of different things as well.